x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/job/persistence/AnomalyDetectorsIndexTests.java

```diff
@@ -0,0 +1,204 @@
+        when(indicesAdminClient.prepareAliases()).thenReturn(new IndicesAliasesRequestBuilder(client, IndicesAliasesAction.INSTANCE));
+        doAnswer(withResponse(new AcknowledgedResponse(true))).when(indicesAdminClient).aliases(any(), any());
+
+        adminClient = mock(AdminClient.class);
+        when(adminClient.indices()).thenReturn(indicesAdminClient);
+
+        client = mock(Client.class);
+        when(client.threadPool()).thenReturn(threadPool);
+        when(client.admin()).thenReturn(adminClient);
+
+        finalListener = mock(ActionListener.class);
+
+        createRequestCaptor = ArgumentCaptor.forClass(CreateIndexRequest.class);
+        aliasesRequestCaptor = ArgumentCaptor.forClass(IndicesAliasesRequest.class);
+    }
+
+    @After
+    public void verifyNoMoreInteractionsWithMocks() {
+        verifyNoMoreInteractions(indicesAdminClient, finalListener);
+    }
+
+    public void testCreateStateIndexAndAliasIfNecessary_CleanState() {
+        ClusterState clusterState = createClusterState(Collections.emptyMap());
+        AnomalyDetectorsIndex.createStateIndexAndAliasIfNecessary(client, clusterState, finalListener);
+
+        InOrder inOrder = inOrder(indicesAdminClient, finalListener);
+        inOrder.verify(indicesAdminClient).prepareCreate(".ml-state");
+        inOrder.verify(indicesAdminClient).create(createRequestCaptor.capture(), any());
+        inOrder.verify(finalListener).onResponse(true);
+
+        CreateIndexRequest createRequest = createRequestCaptor.getValue();
+        assertThat(createRequest.index(), equalTo(".ml-state"));
+        assertThat(createRequest.aliases(), equalTo(Collections.singleton(new Alias(".ml-state-write"))));
+    }
+
+    private void testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExists(String indexName) {
+        ClusterState clusterState = createClusterState(Collections.singletonMap(indexName, createIndexMetaDataWithAlias(indexName)));
+        AnomalyDetectorsIndex.createStateIndexAndAliasIfNecessary(client, clusterState, finalListener);
+
+        InOrder inOrder = inOrder(finalListener);
+        inOrder.verify(finalListener).onResponse(false);
+    }
+
+    public void testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExistsAndPointsAtInitialStateIndex() {
+        testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExists(".ml-state-000001");
+    }
+
+    public void testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExistsAndPointsAtSubsequentStateIndex() {
+        testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExists(".ml-state-000007");
+    }
+
+    public void testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExistsAndPointsAtDummyIndex() {
+        testCreateStateIndexAndAliasIfNecessary_WriteAliasAlreadyExists("dummy-index");
+    }
+
+    private void testCreateStateIndexAndAliasIfNecessary_WriteAliasDoesNotExist(List<String> existingIndexNames,
```

Comment:
nit: its confusing to prefix with `test` as this is usually the entry point (if public) for a unit test. I tend towards naming inner methods like this `assert...`

Reply:
Done.